core/node/l1_recovery/src/l1_fetcher/types/mod.rs

```diff
@@ -0,0 +1,308 @@
+        types::common::{parse_l2_to_l1_log, read_next_n_bytes},
+    },
+    storage::PackingType,
+};
+
+// NOTE: We should probably make these more human-readable.
+pub mod common;
+pub mod v1;
+pub mod v2;
+pub mod v3;
+
+#[allow(clippy::enum_variant_names)]
+#[derive(Error, Debug, Clone)]
+pub enum ParseError {
+    #[error("invalid Calldata: {0}")]
+    InvalidCalldata(String),
+
+    #[error("invalid StoredBlockInfo: {0}")]
+    InvalidStoredBlockInfo(String),
+
+    #[error("invalid CommitBlockInfo: {0}")]
+    InvalidCommitBlockInfo(String),
+
+    #[allow(dead_code)]
+    #[error("invalid compressed bytecode: {0}")]
+    InvalidCompressedByteCode(String),
+
+    #[error("invalid compressed value: {0}")]
+    InvalidCompressedValue(String),
+
+    #[error("invalid pubdata source: {0}")]
+    InvalidPubdataSource(String),
+
+    #[error("blob storage error: {0}")]
+    BlobStorageError(String),
+
+    #[error("blob format error: {0}")]
+    BlobFormatError(String, String),
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+pub enum L2ToL1Pubdata {
+    L2ToL1Log {
+        shard_id: u8,
+        is_service: bool,
+        tx_number_in_block: u16,
+        sender: Address,
+        key: H256,
+        value: H256,
+    },
+    L2ToL2Message(Vec<u8>),
+    PublishedBytecode(Vec<u8>),
+    CompressedStateDiff {
+        is_repeated_write: bool,
+        derived_key: U256,
+        packing_type: PackingType,
+    },
+}
+
+pub trait CommitBlockFormat {
+    fn to_enum_variant(self) -> CommitBlockInfo;
+}
+
+#[derive(Debug)]
+pub enum CommitBlockInfo {
+    V1(V1),
+    V2(V2),
+}
+
+/// Block with all required fields extracted from a [`CommitBlockInfo`].
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct CommitBlock {
+    /// ZKSync batch number.
+    pub l1_batch_number: u64,
+    /// Next unused key serial number.
+    pub index_repeated_storage_changes: u64,
+    /// The state root of the full state tree.
+    pub new_state_root: Vec<u8>,
+    /// Storage write access as a concatenation key-value.
+    pub initial_storage_changes: IndexMap<U256, PackingType>,
+    /// Storage write access as a concatenation index-value.
+    pub repeated_storage_changes: IndexMap<u64, PackingType>,
+    /// (contract bytecodes) array of L2 bytecodes that were deployed.
+    pub factory_deps: Vec<Vec<u8>>,
+
+    pub l1_tx_count: u64,
+
+    pub timestamp: u64,
+    pub priority_operations_hash: H256,
+    pub priority_ops_onchain_data: Vec<PriorityOpOnchainData>,
+    pub l2_logs_tree_root: H256,
+    pub commitment: H256,
+    pub rollup_last_leaf_index: u64,
+}
+
+impl CommitBlock {
+    pub fn try_from_token<'a, F>(value: &'a ethabi::Token) -> Result<Self, ParseError>
+    where
+        F: CommitBlockFormat + TryFrom<&'a ethabi::Token, Error = ParseError>,
+    {
+        let commit_block_info = F::try_from(value)?;
+        Ok(Self::from_commit_block(commit_block_info.to_enum_variant()))
+    }
```

Comment:
wouldn't https://github.com/matter-labs/zksync-era/tree/main/core/lib/l1_contract_interface be a better place for this logic?